File: generate_munis.py
import urllib.request
import urllib.parse
import json
import random

def get_category_members(category):
    category_encoded = urllib.parse.quote(category)
    url = f"https://fr.wikipedia.org/w/api.php?action=query&list=categorymembers&cmtitle=Category:{category_encoded}&cmlimit=500&format=json"
    req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
    with urllib.request.urlopen(req) as response:
        data = json.loads(response.read().decode('utf-8'))
    return [m['title'] for m in data['query']['categorymembers'] if not m['title'].startswith('Catégorie:')]

bruxelles = get_category_members("Commune_dans_la_région_de_Bruxelles-Capitale")

# We want filter by province too!
provinces = {
    "Brabant wallon": "Commune_dans_la_province_du_Brabant_wallon",
    "Hainaut": "Commune_dans_la_province_de_Hainaut",
    "Liège": "Commune_dans_la_province_de_Liège",
    "Luxembourg": "Commune_dans_la_province_de_Luxembourg",
    "Namur": "Commune_dans_la_province_de_Namur"
}

munis = []

for name in bruxelles:
    name = name.split(" (")[0]
    munis.append({'name': name, 'region': 'Bruxelles', 'province': 'Bruxelles'})

for prov_name, category in provinces.items():
    members = get_category_members(category)
    for name in members:
        name = name.split(" (")[0]
        munis.append({'name': name, 'region': 'Wallonie', 'province': prov_name})

# Clean duplicates if any
seen = set()
unique_munis = []
for m in munis:
    if m['name'] not in seen:
        seen.add(m['name'])
        unique_munis.append(m)

with open('src/data/municipalities.ts', 'w', encoding='utf-8') as f:
    f.write("export interface BaseCommune {\n")
    f.write("  name: string;\n")
    f.write("  region: string;\n")
    f.write("  province: string;\n")
    f.write("  ipp: number;\n")
    f.write("  pri: number;\n")
    f.write("  taxCount: number;\n")
    f.write("}\n\n")
    f.write("export const municipalities: BaseCommune[] = [\n")
    for m in unique_munis:
        ipp = round(5 + random.random() * 4, 1)
        pri = int(2000 + random.random() * 1500)
        taxCount = int(10 + random.random() * 50)
        name_escaped = m['name'].replace("'", "\\'")
        f.write(f"  {{ name: '{name_escaped}', region: '{m['region']}', province: '{m['province']}', ipp: {ipp}, pri: {pri}, taxCount: {taxCount} }},\n")
    f.write("];\n")

print(f"Generated {len(unique_munis)} municipalities.")

File: generate_munis.js
import fs from 'fs';

async function fetchCommunes() {
  try {
    // Wikidata query to get municipalities of Wallonia and Brussels
    // Q240 = Brussels
    // Q3392 = Wallonia
    // Q493522 = municipality of Belgium
    const query = `
      SELECT ?muniLabel ?regionLabel WHERE {
        ?muni wdt:P31/wdt:P279* wd:Q493522.
        ?muni wdt:P17 wd:Q31.
        ?muni wdt:P131* ?region.
        FILTER(?region IN (wd:Q240, wd:Q3392))
        SERVICE wikibase:label { bd:serviceParam wikibase:language "fr". }
      }
    `;
    const url = 'https://query.wikidata.org/sparql?query=' + encodeURIComponent(query) + '&format=json';
    const response = await fetch(url, { headers: { 'User-Agent': 'Mozilla/5.0 (Node.js)' }});
    const data = await response.json();
    
    const results = data.results.bindings;
    let munis = new Map();
    
    for (let row of results) {
       const name = row.muniLabel.value;
       const region = row.regionLabel.value;
       if (!name.startsWith('Q')) {
          munis.set(name, region.includes('Bruxelles') ? 'Bruxelles' : 'Wallonie');
       }
    }
    
    const array = Array.from(munis.entries()).map(([name, region]) => {
      // random values for mock data
      const ipp = (5 + Math.random() * 4).toFixed(1);
      const pri = Math.floor(2000 + Math.random() * 1500);
      const taxCount = Math.floor(10 + Math.random() * 50);
      return `  { name: "${name.replace(/"/g, '\\"')}", region: '${region}', ipp: ${ipp}, pri: ${pri}, taxCount: ${taxCount} }`;
    });
    
    const output = `import { BaseCommune } from '../types';

export const municipalities: BaseCommune[] = [
${array.join(',\n')}
];
`;
    fs.writeFileSync('src/data/municipalities.ts', output);
    console.log(`Generated ${array.length} municipalities.`);
  } catch (e) {
    console.error(e);
  }
}
fetchCommunes();
